### ISSUE TITLE
[BUG] LR set in config json does not seem to have any effect

### ISSUE THREAD

**Issue by orcinus** [victorchall/EveryDream2trainer#120]
These are two different runs, overlaid on top of each other - one with LR 4e-6, the other one with LR 7e-6:

![image](https://user-images.githubusercontent.com/1574258/227693576-e24c6fed-2255-415f-a125-1b2c65c64554.png)

I've done another one immeditely after at 1e-6 - exact same loss.
Is the optimizer.json LR overwriting the config file one by any chance?

**Comment by victorchall** [comment]
The actual value used is logged.  Look at your log or attach your log.  

You should see something like this:
```    unet lr: 1e-06, text encoder lr: 5e-07, betas: [0.9, 0.999], epsilon: 1e-08, weight_decay: 0.01 *```

The shape of the loss graph will likely be identical unless you use seed -1.  This is expected. 

You can turn on validation (see docs) and the val/loss is going to be more meaningful to see how your LR is affecting training.

The LR in the main training config json takes precedent over the one in the optimizer.json.  It was a bit safer since that's what people are used to using, but you can literally delete the LR out of the main config if you use optimiser.json if you want.

There's a log line if your LR is overriding.  You will see "Overriding LR from optimizer config with main config/cli LR setting" if the main LR is overriding the one in optimizer.json

Attach your logs and configs if you have more questions.  Can't troubleshoot blind.

**Comment by orcinus** [comment]
> The shape of the loss graph will likely be identical unless you use seed -1. This is expected.

So if i set the LR to 0, it will still be the same as above with the same seed?
That makes absolutely no intuitive sense :)

It's not just the shape that's the same.
The *values* are identical (look at the Y axis).

**Comment by victorchall** [comment]
The loss is not directly affected by LR, only over over time yes, but the normal loss value is influenced significantly by RNG as well and spotting a trend is very difficult. 

Validation loss is MUCH more obvious.  **_Use validation if you care about tracking loss._**  The loss/log_step and loss/epoch are to be honest not very valuable.

Try reading the documents on validation and turning it on.  Then compare your values.

You don't need to use photoshop, use use tensorboard or wandb and leave your logs in tact, it will overlay them.  For W&B make sure you leave the project_name the same so it knows the runs belong together, you can put a value in "run_name" like "1e-6 test" and "4e-6 test" so its easier to compare.

**Comment by orcinus** [comment]
> The loss is not directly affected by LR, only over over time yes

But that's exactly what i'm saying - i was getting the same timebase, and the same Y values, regardless of the LR i've set.
That makes no sense.
I'll update and see if i can replicate it.

**Comment by victorchall** [comment]
Use validation, you should see a clear trend difference in different LR values.  If that shows no change then there is a bug, otherwise the photoshopped overlay of loss/log_step is too hard to interpret.

I would compare something like 5e-7 and 2e-6.   7e-6 is what I'd called extremely high and you may be cooking the model extremely quickly before you can even see the trends.

**Comment by orcinus** [comment]
> Val is only run once per epoch at most, don't know how many epochs that is or how many images you have. It may be easier to test using a reasonable data set, like a few hundred images.

This was 1066 images, batch size 1.

**Comment by victorchall** [comment]
4e-6 or 7e-6 is absurdly high for batch size 1.  You should be 1e-6 maximum, probably closer to 5e-7, maybe even lower.

**Comment by orcinus** [comment]
I know it's absurdly high, the point was to demonstrate it has no visible effect on log_step trend :)
In any case, sorry for dropping this, was meaning to create a repro case, but got distracted by other things (VAE fine-tuning/training).
